modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4WriteThrottlingHandler.java

```diff
@@ -43,7 +46,54 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
         assert msg instanceof ByteBuf;
         assert Transports.assertDefaultThreadContext(threadContext);
         assert Transports.assertTransportThread();
-        final boolean queued = queuedWrites.offer(new WriteOperation((ByteBuf) msg, promise));
+        final ByteBuf buf = (ByteBuf) msg;
+        if (ctx.channel().isWritable() && currentWrite == null && queuedWrites.isEmpty()) {
+            // nothing is queued for writing and the channel is writable, just pass the write down the pipeline directly
+            if (buf.readableBytes() > MAX_BYTES_PER_WRITE) {
+                writeInSlices(ctx, promise, buf);
+            } else {
+                ctx.write(msg, promise);
+            }
+        } else {
+            queueWrite(buf, promise);
+        }
+    }
+
+    /**
+     * Writes slices of up to the max write size until the channel stops being writable or the message has been written in full.
+     */
+    private void writeInSlices(ChannelHandlerContext ctx, ChannelPromise promise, ByteBuf buf) {
+        while (true) {
+            final int readableBytes = buf.readableBytes();
+            final int bufferSize = Math.min(readableBytes, MAX_BYTES_PER_WRITE);
+            if (readableBytes == bufferSize) {
+                // last write for this chunk we're done
+                ctx.write(buf, promise);
+                return;
+            }
+            final int readerIndex = buf.readerIndex();
+            ByteBuf writeBuffer = buf.retainedSlice(readerIndex, bufferSize);
+            buf.readerIndex(readerIndex + bufferSize);
+            ctx.write(writeBuffer).addListener(forwardFailureListener(ctx, promise));
+            if (ctx.channel().isWritable() == false) {
+                // channel isn't writable any longer -> move to queuing
+                queueWrite(buf, promise);
+                return;
+            }
+        }
+    }
+
+    private static GenericFutureListener<Future<? super Void>> forwardFailureListener(ChannelHandlerContext ctx, ChannelPromise promise) {
```

Comment:
Just a note on why the future handling here is different from that in the http handler:

this listener is a little dirty and we do it nicer (but more expensive) in `Netty4HttpPipeliningHandler` where we use a proper promise combiner. We could align this to use a promise combiner as well or use this optimization that does not combine promises in the REST layer as well (it's not really necessary to use the combiner when all the messages are handled in order in all downstream handlers).